File: assets/cs/ws/up-client/templates/services/WSService.cs
using System;
using System.Threading.Tasks;
using System.Collections.Generic;
using nex.ws;

namespace {{namespace}}
{
    public class {{serviceUpperName}}WSService<TUser, TToken>: WSServiceBase<TUser, TToken>
    {
        #region [ implement WSServiceBase ]
        public override string Name { get { return "{{serviceName}}"; } }
        #endregion

        #region [ constructor ]
        public {{serviceUpperName}}WSService(RestClient<TUser, TToken> rest, HubClient<TUser, TToken> hub)
            :base(rest, hub)
        {
            {{#hubEvents}}
            _{{name}} = new {{notification}}{{&arguments}}(hub, Name, "{{name}}");
            {{/hubEvents}}
        }
        #endregion

        #region [ hub private ]
        {{#hubEvents}}
        private {{notification}}{{&arguments}} _{{name}} { get; }
        {{/hubEvents}}
        #endregion

        #region [ hub public ]
        {{#hubEvents}}

        // isAuth: {{isAuth}}
        public {{notification}}{{&arguments}} {{name}} { get { return _{{name}}; } }
        {{/hubEvents}}
        #endregion

        #region [ rest ]
        {{#restMethods}}

        // isAuth: {{isAuth}}
        public Task{{&returnType}} {{name}}({{&methodParams}}) { return Request{{&returnType}}( "{{name}}", {{&requestParams}} ); }
        {{/restMethods}}
        #endregion
    }
}

File: assets/ts/ws/up-client/templates/services/WSService.ts
import { lib } from '../..'

{{#imports}}
import { {{&type}} } from '../../models/{{&path}}'
{{/imports}}

export class {{serviceUpperName}}WSService extends lib.WSServiceBase {
    //#region [ implement WSServiceBase ]
    public readonly name = '{{serviceName}}'
    //#endregion

    //#region [ hub ]
    {{#hubEvents}}

    // isAuth: {{isAuth}}
    public readonly {{name}} = new {{notification}}{{&arguments}}( this._hub, this.name, '{{name}}' )
    {{/hubEvents}}
    //#endregion

    //#region [ rest ]
    {{#restMethods}}

    // isAuth: {{isAuth}}
    {{#methodParams.length}}
    public {{name}} ( {{&methodParams}} ): Promise{{&returnType}} {
        return this.request{{&returnType}}( '{{name}}', {{&requestParams}})
    }
    {{/methodParams.length}}  
    {{^methodParams.length}}
    public {{name}} ( ): Promise{{&returnType}} {
        return this.request{{&returnType}}( '{{name}}', {{&requestParams}})
    }
    {{/methodParams.length}}  
    {{/restMethods}}
    //#endregion
}
